_posts/2020-09-27-spring-boot-12-factor-app.md

```diff
@@ -0,0 +1,169 @@
+```xml
+    <dependency>
+      <groupId>org.springframework.boot</groupId>
+      <artifactId>spring-boot-starter-data-jpa</artifactId>
+    </dependency>
+    <dependency>
+      <groupId>com.h2database</groupId>
+      <artifactId>h2</artifactId>
+      <scope>runtime</scope>
+    </dependency>
+```
+The declarative format allows us to replace the H2 database with any other RDBMS like Oracle or MySQL.
+
+## Build, Release, Run - Leverage Development Workflow for Containers
+> Strictly separate build and run stages
+
+The stages for Build, Release, Run should be kept separate. For Spring Boot Applications, this is easy to achieve with the development workflow for containers. 
+
+The activities in these stages are:
+***Build***: we compile the source code and build the Docker Image
+***Release***: Tag the Image and push to the Registry.
+***Run***:  The Image pushed to the Registry during the release stage is pulled and run as a container instance. 
+ 
+If we are using containers to package and run our application, no application changes are required to adhere to the Twelve-factor app principle.
+
+
+## Processes - Stateless Applications
+> Execute the app as one or more stateless processes
+
+Spring Boot applications execute as a Java process on the host system or inside a container runtime environment like Docker. This principle advocates that the processes should be stateless and share-nothing.  Any data that needs to persist must be stored in a stateful backing service like a database.
+
+This is a shift from the method of using “sticky sessions” in web applications that cache user session data in the memory of the application's process and expecting future requests from the same session to be routed to the same process. Sticky sessions are a violation of twelve-factor. Session state data should be stored outside the application in a datastore that offers time-expiration, such as Memcached or Redis.
+
+## Port Binding - Port Defined as Environment Property
+> Export services via port binding
+
+Port binding is one of the fundamental requirements for microservices to be autonomous and self-contained.
+
+The default web container- Tomcat is embedded in the Spring Boot applications that exports HTTP as a service by binding to a port and listening to incoming requests in that port. a port that is specified by the property  ```server.port```. The default value is 8080. But we can override this value by passing this property as an environment variable. 
+
+
+## Concurrency - Stateless Applications Helps to Scale-Out
+> Scale out via the process model 
+
+Spring Boot applications are stateless. This helps them to scale out by creating more instances to support increasing loads. This is taken care of by container orchestration systems like Kubernetes and Docker Swarm. From an application perspective all state if any needs to be managed outside the application.
```

Comment:
This is a weak explanation for the Concurrency Factor. I would say that Spring Boot doesn't help us much with this Factor. We have to make sure that our app is stateless and thus can be scaled out to many concurrent workers. And we also have to make sure to split up our apps into multiple apps (i.e. microservices) if we want to scale certain processes separately.